### PR TITLE
IGNITE-17553: Add flatMap function to Flow framework

### PR DESCRIPTION
**Pull request by Flaugh24**
https://issues.apache.org/jira/browse/IGNITE-17553

### CODE REVIEW

File: modules/cli/src/test/java/org/apache/ignite/internal/cli/commands/flow/FlowTest.java [hunk 1]
```diff
@@ -269,24 +288,27 @@ void customDecorator() throws IOException {
...
 
-    private static FlowBuilder<Object, String> askQuestion() {
-        return Flows.question("Do you like this?",
-                        List.of(new QuestionAnswer<>("yes"::equals, (a, i) -> 1),
-                                new QuestionAnswer<>("no"::equals, (a, i) -> 2))
-                )
-                .map(String::valueOf);
+    @Test
+    void flatMapInterrupted() {
+        Flows.from("fizz")
```
Comment (PakhomovAlexander) [inline modules/cli/src/test/java/org/apache/ignite/internal/cli/commands/flow/FlowTest.java:302]: I would like to see more complex tests, like:
 ```
Flows.from("fizz")
   .map(v -> Flows.from(it -> it + "1"))
   .print()
   .flatMap(v -> {
                     throw new FlowInterruptException();
                 })
  .print()
  .map(v -> Flows.from(it -> it + "2"))
  .print()
  .start();
```

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/flow/builder/FlowBuilderImpl.java [hunk 2]
```diff
@@ -69,6 +69,11 @@ public <OT> FlowBuilder<I, OT> then(Flow<O, OT> flow) {
         return new FlowBuilderImpl<>(this.flow.composite(flow), exceptionHandlers, decoratorRegistry);
     }
 
+    @Override
+    public <OT> FlowBuilder<I, OT> flatMap(Function<O, FlowBuilder<O, OT>> mapper) {
+        return then(it -> mapper.apply(it.value()).build().start(it));
+    }
+
     @Override
     public <OT> FlowBuilder<I, O> ifThen(Predicate<O> tester, Flow<O, OT> flow) {
```
Comment (PakhomovAlexander) [inline modules/cli/src/main/java/org/apache/ignite/internal/cli/core/flow/builder/FlowBuilderImpl.java:78]: should we deprecate `ifThen` or something?

**Review by PakhomovAlexander — CHANGES_REQUESTED** [review]

File: modules/cli/src/main/java/org/apache/ignite/internal/cli/core/flow/builder/FlowBuilderImpl.java (hunk 2 above)
Comment (Flaugh24) [reply modules/cli/src/main/java/org/apache/ignite/internal/cli/core/flow/builder/FlowBuilderImpl.java:78]: Why? There are places where we use `ifThen`

File: modules/cli/src/test/java/org/apache/ignite/internal/cli/commands/flow/FlowTest.java (hunk 1 above)
Comment (Flaugh24) [reply modules/cli/src/test/java/org/apache/ignite/internal/cli/commands/flow/FlowTest.java:302]: done

**Review by SammyVimes — APPROVED** [review]
LGTM

**Pull request merged**